git/gitattr/tree_test.go

```diff
@@ -0,0 +1,246 @@
+	defer os.Remove(tmp)
+
+	db, err := gitobj.FromFilesystem(tmp, "")
+	require.NoError(t, err)
+	defer db.Close()
+
+	blob, err := db.WriteBlob(gitobj.NewBlobFromBytes([]byte(`
+		*.dat filter=lfs diff=lfs merge=lfs -text
+	`)))
+	require.NoError(t, err)
+
+	child, err := db.WriteTree(&gitobj.Tree{Entries: []*gitobj.TreeEntry{
+		{
+			Name:     ".gitattributes",
+			Oid:      blob,
+			Filemode: 0100644,
+		},
+	}})
+
+	tree, err := New(db, &gitobj.Tree{Entries: []*gitobj.TreeEntry{
+		{
+			Name:     "child",
+			Oid:      child,
+			Filemode: 040000,
+		},
+	}})
+	require.NoError(t, err)
+
+	attrs := tree.Applied("child/foo.dat")
+
+	assert.Len(t, attrs, 4)
+	assert.Equal(t, attrs[0], &Attr{K: "filter", V: "lfs"})
+	assert.Equal(t, attrs[1], &Attr{K: "diff", V: "lfs"})
+	assert.Equal(t, attrs[2], &Attr{K: "merge", V: "lfs"})
+	assert.Equal(t, attrs[3], &Attr{K: "text", V: "false"})
+}
+
+func TestNewDiscoversIndirectChildrenTrees(t *testing.T) {
+	tmp, err := ioutil.TempDir("", "")
+	require.NoError(t, err)
+	defer os.Remove(tmp)
+
+	db, err := gitobj.FromFilesystem(tmp, "")
+	require.NoError(t, err)
+	defer db.Close()
+
+	blob, err := db.WriteBlob(gitobj.NewBlobFromBytes([]byte(`
+		*.dat filter=lfs diff=lfs merge=lfs -text
+	`)))
+	require.NoError(t, err)
+
+	indirect, err := db.WriteTree(&gitobj.Tree{Entries: []*gitobj.TreeEntry{
+		{
+			Name:     ".gitattributes",
+			Oid:      blob,
+			Filemode: 0100644,
+		},
+	}})
+
+	child, err := db.WriteTree(&gitobj.Tree{Entries: []*gitobj.TreeEntry{
+		{
+			Name:     "indirect",
+			Oid:      indirect,
+			Filemode: 040000,
+		},
+	}})
+
+	tree, err := New(db, &gitobj.Tree{Entries: []*gitobj.TreeEntry{
+		{
+			Name:     "child",
+			Oid:      child,
+			Filemode: 040000,
+		},
+	}})
+	require.NoError(t, err)
+
+	attrs := tree.Applied("child/indirect/foo.dat")
+
+	assert.Len(t, attrs, 4)
+	assert.Equal(t, attrs[0], &Attr{K: "filter", V: "lfs"})
+	assert.Equal(t, attrs[1], &Attr{K: "diff", V: "lfs"})
+	assert.Equal(t, attrs[2], &Attr{K: "merge", V: "lfs"})
+	assert.Equal(t, attrs[3], &Attr{K: "text", V: "false"})
+}
+
+func TestNewIgnoresChildrenAppropriately(t *testing.T) {
+	tmp, err := ioutil.TempDir("", "")
+	require.NoError(t, err)
+	defer os.Remove(tmp)
+
+	db, err := gitobj.FromFilesystem(tmp, "")
+	require.NoError(t, err)
+	defer db.Close()
+
+	blob, err := db.WriteBlob(gitobj.NewBlobFromBytes([]byte(`
+		*.dat filter=lfs diff=lfs merge=lfs -text
+	`)))
+	require.NoError(t, err)
+
+	child, err := db.WriteTree(&gitobj.Tree{Entries: []*gitobj.TreeEntry{
+		{
+			Name:     "README.md",
+			Oid:      []byte("00000000000000000000"),
+			Filemode: 0100644,
+		},
+	}})
+
+	tree, err := New(db, &gitobj.Tree{Entries: []*gitobj.TreeEntry{
+		{
+			Name:     ".gitattributes",
+			Oid:      blob,
+			Filemode: 0100644,
+		},
+		{
+			Name:     "child",
+			Oid:      child,
+			Filemode: 040000,
+		},
+	}})
+	require.NoError(t, err)
+
+	assert.NotContains(t, tree.Children, "child")
+}
```

Comment:
It might also be worth adding a test that when parsing a `.gitattributes` file, the `tree` successfully picks up multiple lines of attributes

Reply:
🤔, I think I might recommend against this unless you feel strongly about it. The `New()` function only cares about discovering trees and their .gitattributes, not parsing them. The later is the responsibility of `ParseLines()`, which does have tests ensuring that multi-line `.gitattributes` are read correctly.

I suppose that we could add a test like that that tests the composability of the two, but I'm not sure that given our knowledge of how the code is implemented it would provide worthwhile benefit.